File: Array/minAndMaxUsingDandQ.cpp
/*
    Find min and max using divid and conqor 
*/
#include<iostream>

using namespace std;

int abs(int value){
	if (value < 0)
	{
		return -value;
	}
	return value;
}

int * minAndmax(int *array,int start,int end,int min,int max){
	// If the array has only one element
	if(start == end){
		int results[2] = {min,max};
		return &results;
	}
	else if(abs(start - end) == 1){
		// If it has two elements
		int results[2] = {0,0};
		if(array[start] < array[end]){
			results[0] = array[start];
			results[1] = array[end];
		}else{
			results[1] = array[start];
			results[0] = array[end];
		}
		return &results;
	}else{
		// If there are more than 2 elements 
		int mid = (start+end)/2;
		int left[2] = minAndmax(array,start,mid,min,max);
		int right[2] = minAndmax(array,mid+1,end,min,max);
		int results[0] = {0,0};
		if(left[0] < right[0]){
			results[0] = left[0];
		}else{
			results[0] = right[0];
		} 

		if (left[1] < right[1])
		{
			results[1] = left[1];
		}else{
			results[1] = right[1];
		}

		return &results;
	}

}
int main(){
	int elements[5] = {1,54,23,76,12};
	int *array = &elements;
	int results[2] = minAndmax(array,0,5,1,1);
	cout<<results;
	return 0;
}

File: Stacks , Queues/minAndmaxUsingDandQ.py
import math
#Using divide and conquor

def divide(array,start,end,min1,max1):
	if start == end: # Array with one element
		min1 = max1 = array[start]
	elif (math.fabs(start-end) == 1):
		# When there are only two elements 
		li = []
		if array[start] < array[end]:
			min1 = array[start]
			max1 = array[end]			
		else:
			min1 = array[end]
			max1 = array[start]
		li.append(min1)
		li.append(max1)
		return li		
	else:
		# When there are more than 2 elements
		mid = (start+end)/2
		left = divide(array,start,mid,min1,max1)
		right = divide(array,mid+1,end,min1,max1)
		if left is not None:
			if min1 > left[0]:
				min1 = left[0]
			if max1 < left[1]:
				max1 = left[1]
			print "Left min "+str(min1)+" max "+str(max1)
		if right is not None:
			if min1 > right[0]:
				min1 = right[0]
			if max1 < right[1]:
				max1 = right[1]
	

# Without divide and conquor 
def minAndmax(li):
	maxE = minE = li[0]
	length = len(li)
	for i in range(1,length):
		if minE > li[i]:
			minE = li[i]
		elif maxE < li[i]:
			maxE = li[i]
	print "Min "+str(minE)+" Max "+str(maxE)

elements = [1,234,23,641,767,12]

divide(elements,0,len(elements)-1,elements[0],elements[0])